pkg/acs/calacs/acsccd/blev_funcs_postsm4.c

```diff
@@ -36,152 +39,218 @@ static int make_amp_array(const int arr_rows, const int arr_cols, SingleGroup *
-    } else {
-      make_amp_array(arr_rows, arr_cols, chip2, i, ampdata);
-    }
-       
-    /* calculate "magic square mean" */
-    sum = 0.0;
-    num = 0;
+  addPtr(&ptrReg, dc_bias_levels, &free);
     
-    for (j = 2057; j <= 2066; j++) {
-      for (k = 13; k <= 22; k++) {
-        sum += ampdata[arr_cols*j + k];
-        num++;
+  /* array for amp data and amp data + gap pixels */
+  double *ampdata = malloc(arr_rows * arr_cols * sizeof(double));
+  addPtr(&ptrReg, ampdata, &free);
+  double *ampdata_gap = malloc(nquad_pix * sizeof(double));
+  addPtr(&ptrReg, ampdata_gap, &free);
+  
+  /* For full frame data, there are four amps in use.  For supported subarray
+   * data, there is only one amp in use.  The actual amp in use must be determined. 
+   */
+  int numAmpsInUse = (acs->subarray == YES) ? 1 : NAMPS;
+  
+  /* Loop over all the amps for full frame data (total of 4 amps, 0 to 3) or just use 
+   * the single amp value corresponding to the supported subarray (1 amp, value defined in acs.h).
+   */
+  int ampInUse = 0;
+  {unsigned int i;
+  for (i = 0; i < numAmpsInUse; i++) {
+      ampInUse = i;
+      if ((i == 0) && (acs->subarray == YES)) {
+         if (strcmp (acs->ccdamp, "A") == 0)
+            ampInUse = AMP_A;
+         else if (strcmp (acs->ccdamp, "B") == 0)
+            ampInUse = AMP_B;
+         else if (strcmp (acs->ccdamp, "C") == 0)
+            ampInUse = AMP_C;
+         else if (strcmp (acs->ccdamp, "D") == 0)
+            ampInUse = AMP_D;
+      } 
+
+      /* put a single amp's data into ampdata */
+      if (acs->subarray == YES) {
+         make_amp_array(arr_rows, arr_cols, sg[0], ampInUse, ampdata);
+      } else {
+         if (i < 2) {
+            make_amp_array(arr_rows, arr_cols, sg[1], ampInUse, ampdata);
+         } else {
+            make_amp_array(arr_rows, arr_cols, sg[0], ampInUse, ampdata);
+         }
       }
-    }
+       
+      /* calculate "magic square mean" - this works for any single amp */
+      double sum = 0.0;
+      int num = 0;
+      double magic_square_mean = 0.0;
     
-    magic_square_mean = sum / (double) num;
+      {unsigned int j, k;
+      for (j = 2057; j <= 2066; j++) {
+          for (k = 13; k <= 22; k++) {
+              sum += ampdata[arr_cols*j + k];
+              num++;
+          }
+      }}
     
-    /* make amp + gap array */
-    for (j = 0; j < arr_rows; j++) {
-      for (k = 0; k < (arr_cols + ngap_pix); k++) {
-        if (k < arr_cols) {
-          ampdata_gap[(arr_cols + ngap_pix)*j + k] = ampdata[arr_cols*j + k];
-        } else {
-          ampdata_gap[(arr_cols + ngap_pix)*j + k] = magic_square_mean;
+      magic_square_mean = sum / (double) num;
+    
+      /* make amp + gap array */
+      {unsigned int j, k;
+      for (j = 0; j < arr_rows; j++) {
+          for (k = 0; k < (arr_cols + ngap_pix); k++) {
+              if (k < arr_cols) {
+                 ampdata_gap[(arr_cols + ngap_pix)*j + k] = ampdata[arr_cols*j + k];
```

Comment:
I believe the preference (within HSTCAL) is to use the macro ``Pix(a,i,j)`` ([src](https://github.com/spacetelescope/hstcal/blob/master/include/hstio.h#L188-L196)) for readability, just make sure that when it's expanded you have the order correct.

Reply:
This scheme is used in the acsccd and acscte modules where the specific WFC amp has been remapped into a 1D array.  The knowledge of the location of the physical and virtual overscan regions, as well as the use of the Pix() structure, is in the remapping routine.  In essence, regardless of the actual layout of the specific amp, the remapping into this 1D array allows all the amps to be treated in the same fashion by this routine.  Another motivator is that the coding of this routine is very closely based upon a description laid out in [ISR ACS 2012-02 (http://www.stsci.edu/hst/acs/documents/isrs/isr1202.pdf)

Reply:
Breadcrumb: They're never _not_ 1D arrays, it's C. The re-mapping is only the rotation of the "amps" such that the position of the amplifiers all have the same logical position within the above array. This has nothing to do with my style comment about not having "naked" array equations.